src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp

```diff
@@ -77,62 +103,70 @@ void WriteBufferFromAzureBlobStorage::execWithRetry(std::function<void()> func,
+    /// then we use single part upload instead of multi part upload
+    if (buffer_allocation_policy->getBufferNumber() == 1)
+    {
+        size_t data_size = size_t(position() - memory.data());
+        if (data_size <= max_single_part_upload_size)
+        {
+            Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(memory.data()), data_size);
+            execWithRetry([&](){ block_blob_client.Upload(memory_stream); }, max_unexpected_write_error_retries, data_size);
+            LOG_TRACE(log, "Committed single block for blob `{}`", blob_path);
+            return;
+        }
+    }
+
+
     execWithRetry([this](){ next(); }, max_unexpected_write_error_retries);
 
-    if (tmp_buffer_write_offset > 0)
-        uploadBlock(tmp_buffer->data(), tmp_buffer_write_offset);
+    task_tracker->waitAll();
 
-    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
     execWithRetry([&](){ block_blob_client.CommitBlockList(block_ids); }, max_unexpected_write_error_retries);
 
     LOG_TRACE(log, "Committed {} blocks for blob `{}`", block_ids.size(), blob_path);
 }
 
-void WriteBufferFromAzureBlobStorage::uploadBlock(const char * data, size_t size)
+void WriteBufferFromAzureBlobStorage::nextImpl()
 {
-    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
-    const std::string & block_id = block_ids.emplace_back(getRandomASCIIString(64));
-
-    Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(data), size);
-    execWithRetry([&](){ block_blob_client.StageBlock(block_id, memory_stream); }, max_unexpected_write_error_retries, size);
-    tmp_buffer_write_offset = 0;
-
-    LOG_TRACE(log, "Staged block (id: {}) of size {} (blob path: {}).", block_id, size, blob_path);
+    task_tracker->waitIfAny();
+    writePart();
+    allocateBuffer();
 }
 
-WriteBufferFromAzureBlobStorage::MemoryBufferPtr WriteBufferFromAzureBlobStorage::allocateBuffer() const
+void WriteBufferFromAzureBlobStorage::allocateBuffer()
 {
-    return std::make_unique<Memory<>>(max_single_part_upload_size);
+    buffer_allocation_policy->nextBuffer();
+    auto size = buffer_allocation_policy->getBufferSize();
+    memory = Memory(size);
+    WriteBuffer::set(memory.data(), memory.size());
 }
 
-void WriteBufferFromAzureBlobStorage::nextImpl()
+void WriteBufferFromAzureBlobStorage::writePart()
 {
-    size_t size_to_upload = offset();
+    std::shared_ptr<PartData> part_data;
+    auto data_size = size_t(position() - memory.data());
+    part_data = std::make_shared<PartData>(std::move(memory), data_size);
+    WriteBuffer::set(nullptr, 0);
 
-    if (size_to_upload == 0)
+    if (part_data->data_size == 0)
         return;
 
-    if (!tmp_buffer)
-        tmp_buffer = allocateBuffer();
-
-    size_t uploaded_size = 0;
-    while (uploaded_size != size_to_upload)
+    auto upload_worker = [&, part_data] ()
```

Comment:
Do you really need `&` here inside `[ ]`? It only makes finding mistakes more difficult.

Reply:
I changed it to `= [this, part_data]`, we need `this` for  `blob_container_client`, `write_settings` & `max_unexpected_write_error_retries`